File: examples/typescript/evm/getSwapQuote.ts
// Usage: pnpm tsx evm/getSwapQuote.ts

import { CdpClient } from "@coinbase/cdp-sdk";
import { formatUnits, parseEther } from "viem";
import "dotenv/config";

const cdp = new CdpClient();

// Token definitions for the example
const TOKENS = {
  WETH: {
    address: "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2",
    symbol: "WETH",
    decimals: 18,
  },
  USDC: {
    address: "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
    symbol: "USDC",
    decimals: 6,
  },
};

async function main() {
  // Get or create an account to use for the swap
  const account = await cdp.evm.getOrCreateAccount({ name: "SwapAccount" });
  console.log(`Using account: ${account.address}`);

  try {
    // Define the tokens we're working with
    const sellToken = TOKENS.WETH;
    const buyToken = TOKENS.USDC;
    
    // Get a quote for swapping WETH to USDC on Ethereum mainnet
    const swapQuote = await cdp.evm.getSwapQuote({
      network: "ethereum",
      buyToken: buyToken.address,
      sellToken: sellToken.address,
      sellAmount: parseEther("0.1"), // 0.1 WETH in wei
      taker: account.address,
    });

    // Check for liquidity
    if (!swapQuote.liquidityAvailable) {
      console.log("Insufficient liquidity available for this swap.");
      return;
    }

    // At this point we know liquidityAvailable is true and we can access other properties
    console.log("Swap Quote:");
    console.log("-------------");
    console.log(`Buy Amount: ${formatUnits(swapQuote.buyAmount, buyToken.decimals)} ${buyToken.symbol}`);
    console.log(`Min Buy Amount: ${formatUnits(swapQuote.minBuyAmount, buyToken.decimals)} ${buyToken.symbol}`);
    console.log(`Sell Amount: ${formatUnits(swapQuote.sellAmount, sellToken.decimals)} ${sellToken.symbol}`);
    console.log(`Gas Estimate: ${swapQuote.gas}`);
    
    // Calculate and display price ratios
    const sellAmountBigInt = BigInt(swapQuote.sellAmount);
    const buyAmountBigInt = BigInt(swapQuote.buyAmount);
    
    // Calculate price: How many buyTokens per sellToken
    const sellTokenPrice = Number(
      (buyAmountBigInt * BigInt(10 ** (18 - buyToken.decimals))) / 
      (sellAmountBigInt * BigInt(10 ** (18 - sellToken.decimals)))
    ) / (10 ** 18);
    
    // Calculate inverse price: How many sellTokens per buyToken
    const buyTokenPrice = Number(
      (sellAmountBigInt * BigInt(10 ** (18 - sellToken.decimals))) / 
      (buyAmountBigInt * BigInt(10 ** (18 - buyToken.decimals)))
    ) / (10 ** 18);
    
    console.log("\nToken Price Calculations:");
    console.log("------------------------");
    console.log(`1 ${sellToken.symbol} = ${sellTokenPrice.toLocaleString()} ${buyToken.symbol}`);
    console.log(`1 ${buyToken.symbol} = ${buyTokenPrice.toLocaleString()} ${sellToken.symbol}`);
    
    // Check for any issues
    if (swapQuote.issues.allowance) {
      console.log("\nAllowance Issues:");
      console.log(`Current Allowance: ${swapQuote.issues.allowance.currentAllowance}`);
      console.log(`Spender: ${swapQuote.issues.allowance.spender}`);
    }
    
    if (swapQuote.issues.balance) {
      console.log("\nBalance Issues:");
      console.log(`Current Balance: ${swapQuote.issues.balance.currentBalance}`);
      console.log(`Required Balance: ${swapQuote.issues.balance.requiredBalance}`);
    }

    if (swapQuote.issues.simulationIncomplete) {
      console.log("\n⚠️ WARNING: Simulation incomplete. Results may not be accurate.");
    }
  } catch (error) {
    console.error("Error getting swap quote:", error);
  }
}

main().catch(console.error); 

File: examples/python/evm/get_swap_quote.py
"""Example showing how to get swap quotes."""

import asyncio
from cdp import Cdp


async def main():
    """Get swap quotes example."""
    # Configure CDP SDK
    cdp = Cdp.from_json("~/.config/coinbase_cloud/config.json")

    # Define token addresses
    eth_address = "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE"  # ETH (EIP-7528)
    usdc_base = "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913"  # USDC on Base
    usdc_eth = "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48"  # USDC on Ethereum
    weth_eth = "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2"  # WETH on Ethereum

    # Example 1: ETH to USDC quote on Base
    print("Getting ETH to USDC quote on Base...")
    quote = await cdp.evm.get_quote(
        from_token=eth_address,
        to_token=usdc_base,
        amount="1000000000000000000",  # 1 ETH
        network="base",
    )
    
    print(f"Quote ID: {quote.quote_id}")
    print(f"From: {quote.from_amount} (ETH)")
    print(f"To: {quote.to_amount} (USDC)")
    print(f"Price ratio: {quote.price_ratio}")
    print(f"Expires at: {quote.expires_at}")
    
    # Example 2: USDC to ETH on Base
    print("\n" + "="*50)
    print("Getting USDC to ETH quote on Base...")
    
    quote2 = await cdp.evm.get_quote(
        from_token=usdc_base,
        to_token=eth_address,
        amount="2000000000",  # 2000 USDC
        network="base",
    )
    
    print(f"Quote ID: {quote2.quote_id}")
    print(f"From: {quote2.from_amount} (USDC)")
    print(f"To: {quote2.to_amount} (ETH)")
    
    # Example 3: ERC20 to ERC20 on Ethereum
    print("\n" + "="*50)
    print("Getting USDC to WETH quote on Ethereum...")
    
    quote3 = await cdp.evm.get_quote(
        from_token=usdc_eth,
        to_token=weth_eth,
        amount="1000000000",  # 1000 USDC
        network="ethereum",
    )
    
    print(f"Quote ID: {quote3.quote_id}")
    print(f"From: {quote3.from_amount} (USDC)")
    print(f"To: {quote3.to_amount} (WETH)")
    print(f"Price ratio: {quote3.price_ratio}")


if __name__ == "__main__":
    asyncio.run(main()) 
